src/service_daemon.rs

```diff
@@ -569,6 +659,47 @@ impl ServiceDaemon {
                 zc.add_retransmission(next_time, Command::Browse(ty, delay, listener));
             }
 
+            Command::ResolveHostname(hostname, next_delay, listener, timeout) => {
+                let addr_list: Vec<_> = zc.intf_socks.keys().collect();
+                if let Err(e) = listener.send(ResolutionEvent::SearchStarted(format!(
+                    "{} on addrs {:?}",
+                    &hostname, &addr_list
+                ))) {
+                    error!(
+                        "Failed to send ResolveStarted({})(repeating:{}): {}",
+                        &hostname, repeating, e
+                    );
+                    return;
+                }
+                if !repeating {
+                    zc.add_hostname_resolver(hostname.to_owned(), listener.clone(), timeout);
+                    // if we already have the records in our cache, just send them
+                    zc.query_cache_for_hostname(&hostname, listener.clone());
+                }
+
+                zc.send_query(&hostname, TYPE_A);
+                zc.send_query(&hostname, TYPE_AAAA);
+                zc.increase_counter(Counter::ResolveHostname, 1);
+
+                let now = current_time_millis();
+                let next_time = now + u64::from(next_delay) * 1000;
+                let max_delay = 60 * 60;
+                let delay = cmp::min(next_delay * 2, max_delay);
+
+                // Only add retransmission if it does not exceed the hostname resolver timeout, if any.
+                if zc
+                    .hostname_resolver_timeouts
+                    .get(&hostname)
+                    .map(|timeout| next_time < *timeout)
+                    .unwrap_or(true)
+                {
+                    zc.add_retransmission(
+                        next_time,
+                        Command::ResolveHostname(hostname, delay, listener, timeout),
```

Comment:
I think this already does what you're describing. `add_hostname_resolver` is only called on the first instance of the command due to the repeating flag, so the timeout will be from the moment the caller issue the command.

But it's not very transparent to the reader that all of this happens in `add_hostname_resolver`. Should I move it to another function (`add_hostname_resolver_timeout` or something), and call it below `add_hostname_resolver` in the non-repeat block? I can also set `timeout` to `None` in the retransmissions since the data is no longer needed, and to avoid further confusion.

Reply:
I've set the timeout to `None` in the retransmissions, but kept the rest in `add_hostname_resolver`. Let me know if you would like me to change it

Reply:
👍  this is good